internal/client/db/pinned_blocks_cache.go

```diff
@@ -0,0 +1,138 @@
+// Copyright 2024 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package db
+
+import (
+	"log"
+	"math"
+
+	"github.com/ChainSafe/gossamer/internal/primitives/runtime"
+	lru "github.com/hashicorp/golang-lru/v2"
+)
+
+// Entry for pinned blocks cache.
+type pinnedBlocksCacheEntry struct {
+	// How many times this item has been pinned
+	refCount uint32
+
+	// Cached justifications for this block
+	Justifications runtime.Justifications
+
+	// Cached body for this block
+	Body *[]runtime.Extrinsic
+}
+
+func (pbce *pinnedBlocksCacheEntry) DecreaseRef() {
+	if pbce.refCount > 0 {
+		pbce.refCount--
+	}
+}
+
+func (pbce *pinnedBlocksCacheEntry) IncreaseRef() {
+	if pbce.refCount < math.MaxUint32 {
```

Comment:
Should we throw an error or something if this isn't met since then we loose track of the correct number of references?

Reply:
Good catch!  I think we should panic in this case.